### PR TITLE
RNN Classifiers

### PR DESCRIPTION
**Pull request by tblah**
GRU, LSTM, and BasicRNN.

Roughly 68% accuracy on test data with perfect accuracy on training data so we can probably get something better with tweaks to reduce over-fitting.

### CODE REVIEW

File: Predictors/rnn.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    def __init__(self, CellType=tf.contrib.rnn.GRUCell, n_neurons=100, learning_rate=0.001, n_outputs=4,
+                 n_out_neurons=10, n_inputs=50, n_steps=5, n_epochs=100, batch_size=100, **cell_args):
+        """
+        n_inputs = word encoding size
+        n_steps = amount of recursion e.g. window size
+        n_neurons = number of neurons in RNN
+        n_out_neurons = number of neurons in the fully connected output layer (= the number of classes)
+        n_outputs = number of outputs
+        """
+        self.CellType = CellType
```
Comment (Adimote) [inline Predictors/rnn.py]: instance variables should be snake case (I don't see this being required in the inherited classes), also I notice CellType doesn't feature in the docstring above

File: Predictors/rnn.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+        self.construct_()
+
+    def construct_(self):
+        """
+        tensorflow construction stage
+        """
+
+        self.X = tf.placeholder(tf.float32, [None, self.n_steps, self.n_inputs])
+        # print("\n\nExpecting X.shape = {}, dtype = {}\n".format(self.X.shape, self.X.dtype))
```
Comment (Adimote) [inline Predictors/rnn.py]: in general it's a bad idea to leave commented out debug code, as it's likely to rot. I suggest using python's `logging` library, or just removing it.

File: Predictors/rnn.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        # y = tf.placeholder(tf.int64, [None, self.n_outputs])
```
Comment (Adimote) [inline Predictors/rnn.py]: is this commented out code needed in the future? If so, annotate what it is, otherwise I think it should be removed.

File: Predictors/rnn.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+        sess.run(self.init)
+        for epoch in range(self.n_epochs):
+            try:
+                for iteration in range(sentences.shape[0] // self.batch_size):
+                    # pick a random batch
+                    batch_choices = np.random.choice(X_train.shape[0], self.batch_size, replace=True)
+                    X_batch = X_train[batch_choices]
+                    y_batch = y_train[batch_choices]
+                    # print("type(X_batch = {}, type(y_batch) = {}\n".format(type(X_batch), type(y_batch)))
```
Comment (Adimote) [inline Predictors/rnn.py]: see above about commented out debug code

File: Predictors/rnn.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+
+if __name__ == '__main__':
+    from data.numbered_authors import NumberAuthorsTransformer
+    from data.padded_sentences import PaddedSentenceTransformer
+    from data.import_data import import_data
+    from Predictors.sklearnclassifier import show_stats
+
+    tr, te = import_data()
+    # data_enc = WindowedSentenceTransformer(encoder_size=50)
```
Comment (Adimote) [inline Predictors/rnn.py]: can we put WindowedSentenceTransformer and PaddedSentenceTransformer in an `if` statement? It's better than commenting and uncommenting this test code. It also allows much easier automation later.

File: Predictors/rnn.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+    # for PaddedSentenceTransformer
+    X_train = data_enc.fit_transform(tr.text)
+    X_test = data_enc.transform(te.text)
+
+    sess = tf.Session()
+    with tf.Session() as sess:
+        with sess.as_default():
+            # about 40% accuracy. These two are for WindowedSentenceTransformer
+            # rnn = RNNClassifier(CellType = tf.contrib.rnn.LSTMCell, n_out_neurons=3, n_outputs=1, n_epochs=200, n_neurons=200,
```
Comment (Adimote) [inline Predictors/rnn.py]: probably a good idea to put these commented lines in an `if` statement, too.

**Review by Adimote — APPROVED** [review]
It's good, I have a few comments. 

Having said this, this is much higher quality code than the majority of the code base (there's actual comments!), so i'm approving it.

File: Predictors/rnn.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+            except KeyboardInterrupt:  # stop early with Control+C (SIGINT)
+                print("\nInterrupted by user at epoch {}/{}".format(epoch, self.n_epochs))
+                return self
+
+        return self
+
+    def predict_proba(self, X, sess=None):
+        """
+        probabillites for each class
```
Comment (Adimote) [inline Predictors/rnn.py]: typo

File: Predictors/rnn.py (hunk 1 above)
Comment (tblah) [reply Predictors/rnn.py]: CellType is a class rather than an instance. Is there a way to make that distinction or should it just be named like everything else?

Comment (Adimote) [reply Predictors/rnn.py]: the type of the variable shouldn't really define which naming convention it uses. CapWords is reserved for actual names of classes.

**Comment by tblah** [comment]
Thanks Adimote, I think that is everything. Sorry I really should have known better than not to comment anything.

**Comment by Adimote** [comment]
Looks perfect now, if I could approve twice, I would. feel free to merge, (feel free to ignore those last few linting errors)

**Review by Adimote — APPROVED** [review]
Turns out I can approve twice!
